inst/chunks/fun_xbetax.stan

```diff
@@ -0,0 +1,32 @@
+  /* Extended support beta log-PDF of a single response
+   * Args:
+   *   y: the response value
+   *   mu: mean parameter of the beta distribution
+   *   phi: precision parameter of the beta distribution
+   *   u: the exceedance parameter
+   * Returns:
+   *   a scalar to be added to the log posterior
+   */
+   real xbetax_lpdf(real y, real mu, real phi, real u) {
```

Comment:
can these functions also be vectorized and if yes, would that speed of the sampling speed? The latest versions of Stan allow function overloading also for user-defined functions so you can define the same function multiple times just with different argument signatures.

It is totally alright not to have a vectorized version. I just wanted to ask to be sure.

Reply:
Happy to give that a shot and run some comparisons. Do you know of any vectorized implementation already in **brms** or elsewhere that I can look at?

Reply:
Looking at it again, I think we do not need a vectorized version in this case. I saw you already started and I am sorry I caused you extra work. The reason is the getting speed improvement out of the vectorized version requires a bit more effort since we need to group the y into those that fall into the different categories defined by the if statements. Kind of the same thing as for zero-inflated models, which are also not vectorized for now. In order to safe your time, I would say that the non-vectorized version alone is good enough.

Reply:
No worries. I'll give it another shot and if all fails we default to the scalar version.

Reply:
okay! to be clear: we need the scalar version in any case. an additional vectoriced version may just be helpful to speed up sampling. but this will only be beneficial if the lpdf statements are victorized that is called at the same time for multiple y.

Reply:
Sure. Just for info. How is the choice between vectorized and scalar being made internally? I currently, simply have

```r
stan_log_lik_xbetax <- function(bterms, ...) {
    p <- stan_log_lik_dpars(bterms, reqn = TRUE)
    sdist("xbetax", p$mu, p$phi, p$kappa, vec = FALSE)
}
```

for scalar and set `reqn = FALSE` and `vec = TRUE` for vectorized. There must be a more direct way, from the interface, perhaps?

Reply:
Suppose you have build a vectorized version vectorized only over mu but not over phi and kappa. Then you `reqn = TRUE` only if phi or kappa is predicted or some addition terms are used (like weights), which prevents vectorization from being valid. At the same time, vec = TRUE is only valid if you only predict mu (and keep phi and kappy constant) because you have only build a vectorized version for this case. 

Again, I believe my suggestion to make a vectorized version was not a good once since your family would likely require a bit of work to benefit from vectorization.

Reply:
I think I got a vectorized version in all mu, phi, kappa in commit [31193c8](https://github.com/paul-buerkner/brms/pull/1698/commits/31193c867a6368c2fa2b8b9c34e49a26522a74f3).

Would
```
stan_log_lik_xbetax <- function(bterms, ...) {
    p <- stan_log_lik_dpars(bterms, reqn = FALSE)
    sdist("xbetax", p$mu, p$phi, p$kappa, vec = TRUE)
}
```
be enough to use it? I see

```
R> stan_log_lik_dpars(bterms, reqn = FALSE)
$mu
[1] "mu"

$phi
[1] "phi"

$kappa
[1] "kappa"
```

and 

```
sdist("xbetax", p$mu, p$phi, p$kappa, vec = TRUE)
$dist
[1] "xbetax"

$args
[1] "mu, phi, kappa"

$vec
[1] TRUE

$shift
[1] ""

attr(,"class")
[1] "sdist"
```

and in the output of `stancode()` we have `target += xbetax_lpdf(Y | mu, phi, kappa)`

Does that do it?

P.S.: I might need to check for only ones, only zeros or both ones and zeros but that's easy to do now

Reply:
I think the vectorized version looks good. Now the problem is that this will only work if all 3 parameters are predicted, which will often not be the case. So I would argue you vectorieze (only) the most common case which is, I assume only mu being predicted. Can you check if that leads to relevant speed-ups? 

I will fix the reqn myself before I merge.

Reply:
Wit all three parameters predicted, and a single run, I get:

Vectorized:
```
Chain 1:  Elapsed Time: 232.716 seconds (Warm-up)
Chain 1:                50.756 seconds (Sampling)
Chain 1:                283.472 seconds (Total)
```

Scalar:
```
Chain 1:  Elapsed Time: 253.758 seconds (Warm-up)
Chain 1:                108.226 seconds (Sampling)
Chain 1:                361.984 seconds (Total)
```

which looks good. I will experiment with only one or two parameters being predicted, and report back

Reply:
`inst/chunks/fun_xbetax.stan` now has all eight versions of `xbeta_lpdf`, depending on whether `mu`, `phi`, `kappa` are real or vector. So, given the right call, STAN should figure out which one to use. 

What I could not get my head around from the limited time I had to experiment is how **brms** decides which version to use.  For some reason, the default `stan_log_lik_adj(bterms)` did not do it, but I admit I did not dig further.

For now, I forced the use of the non-vectorized version in the commit for now, having

```
stan_log_lik_xbetax <- function(bterms, ...) {
    p <- stan_log_lik_dpars(bterms, reqn = TRUE)
    sdist("xbetax", p$mu, p$phi, p$kappa, vec = FALSE)
}
```
See, https://github.com/paul-buerkner/brms/pull/1698/commits/6524911fec13e06162e3f9c00b72d9e4324665c7

I'm looking forward to seeing how that choice can be made automatically, when you merge.